File: solutions/python/binary-search/koko-eating-bananas.py
# LeetCode 875 — Koko Eating Bananas (Medium)
# Category: Binary Search · Approach: Search on Answer
# Time: O(n log m) | Space: O(1)
# Source: https://leetcode.com/problems/koko-eating-bananas/

import math

def min_eating_speed(piles: list[int], h: int) -> int:
    lo, hi = 1, max(piles)
    while lo < hi:
        mid = (lo + hi) // 2
        hours = sum(math.ceil(p / mid) for p in piles)
        if hours <= h:
            hi = mid
        else:
            lo = mid + 1
    return lo

File: solutions/typescript/binary-search/koko-eating-bananas.ts
// LeetCode 875 — Koko Eating Bananas (Medium)
// Category: Binary Search · Approach: Search on Answer
// Time: O(n log m) | Space: O(1)
// Source: https://leetcode.com/problems/koko-eating-bananas/

function minEatingSpeed(piles: number[], h: number): number {
  let lo = 1, hi = Math.max(...piles);
  while (lo < hi) {
    const mid = (lo + hi) >> 1;
    const hours = piles.reduce((s, p) => s + Math.ceil(p / mid), 0);
    if (hours <= h) hi = mid;
    else lo = mid + 1;
  }
  return lo;
}
